First iteration of SNN module restructuring

Jegp opened the pull request:
This PR attempts to attack a restructuring of the module interfaces (#166). It is flawed in a number of ways:

* Inconsistency between *Layer, *FF, *Cell modules
* A lot of code duplication
* Divergent code means difficulty to optimize

I have tried to
* Reduce the module interfaces to **three base classes**
  1. `FeedforwardSNN` which represents **shapeless** (no input/hidden shape) and **timeless** feedforward dynamics (also fixed the `FeedForward` misspelling)
  2. `RecurrentSNNCell` which represents **recurrent** and **timeless** dynamics (equivalent to PyTorch's `RNNCell`)
  3. `RecurrentSNN` which represents **recurrent** and **timed** dynamics (equivalent to PyTorch's `RNNBase`/`RNN`)

I would still like to
-  ~[ ] the state to a single tensor instead of a `NamedTuple`~
-  ~[ ] Add the parameters as modules (as suggested by @Huizerd)~
- [x] Migrate LSNN
- [x] Migrate LIF variants (Ex, AdEx, etc.)

What do you think so far? Contributions are more than welcome! :-)

Huizerd commented on norse/torch/module/snn.py:
```diff
@@ -0,0 +1,201 @@
+        T = input_tensor.shape[0]
+        outputs = []
+
+        for ts in range(T):
```
I wondered about this the other day: you can either have a time loop per neuron/layer (as done here), or you can have only one for the entire network. In the former you would go through all time steps for a single layer and then move on to the next and go through all steps for that one etc.; in the latter you would only have one time loop, but for each iteration you would go through all layers.

I was wondering whether there would be a different outcome for these?

Jegp replied:
I think we should facilitate both. And theoretically, the outcome should be the same. The first case is exactly taken care of by the above, but to do the second case, you'd need something that groups all the dynamics into a single module and runs them over all the timesteps. I think this is pretty much what @cpehle tried to do in the [Lift module](https://github.com/norse/norse/blob/master/norse/torch/module/lift.py#L57).

The question we should be asking ourselves, I guess, is whether there is a better way of achieving this. Should we, for instance, augment the [SequentialState](https://github.com/norse/norse/blob/master/norse/torch/module/sequential.py) to also support time? Can we think of something better? Or should we not include these abstractions at all and let people code this themselves?

Huizerd commented on norse/torch/module/snn.py:
```diff
@@ -0,0 +1,201 @@
+        self,
+        activation: FeedforwardActivation,
+        state_fallback: Callable[[torch.Tensor], torch.Tensor],
+        p: torch.nn.Module,
```
I think this would allow the learnable parameters yes

Huizerd commented on norse/torch/module/snn.py:
```diff
@@ -0,0 +1,201 @@
+]
+
+
+class FeedforwardSNN(torch.nn.Module):
```
Not sure about the name, but that's probably for later. It's more of a `Cell` as well, I think

Jegp replied:
@cpehle also just had the idea that we could technically fuse the FF modules with the Cell modules if, for instance, allowed to set the weights to `None`. 

One option could also be to remove the non-temporal recurrent layer, so we could simply have an `SNNCell` and an `SNN` module. That would be much more in line with the PyTorch nomenclature. But I wonder if someone would be missing the non-temporal recurrent dynamics. For instance, in the case of running a sequence of modules for every timestep.

Huizerd commented on norse/torch/module/snn.py:
```diff
@@ -0,0 +1,201 @@
+        p: torch.nn.Module,
+        input_weights: Optional[torch.Tensor] = None,
+        recurrent_weights: Optional[torch.Tensor] = None,
+        autopses: bool = False,
```
Shouldn't this be autapse?

Jegp replied:
Don't know. We also write weight**s**. Theoretically, it is about killing *all* the autopse**s** between all the neurons in the module. But would it be easier to understand in singular?

Jegp replied:
Ah, you meant the spelling :laughing:! Yes, that completely true! But would you also do singular?

Huizerd replied:
No plural is fine, I just meant the 'a' :smile:

Huizerd commented on norse/torch/module/snn.py:
```diff
@@ -0,0 +1,201 @@
+            self.recurrent_weights = torch.randn(hidden_size, hidden_size) * torch.sqrt(
+                2 / hidden_size
+            )
+        self.recurrent_weights = torch.nn.Parameter(
```
I think you can do:
```python
        if recurrent_weights is not None:
            self.recurrent_weights = recurrent_weights
        else:
            self.recurrent_weights = nn.Parameter(
                torch.randn(hidden_size, hidden_size) * torch.sqrt(2 / hidden_size)
            )

        if not autapses:
            with torch.no_grad():
                self.recurrent_weights.fill_diagonal_(0.0)
```

Jegp replied:
Awesome comment! That pretty much removes ~30LOC. Thanks, @Huizerd!

Huizerd commented on norse/torch/module/snn.py:
```diff
@@ -0,0 +1,201 @@
+        input_size: int,
+        hidden_size: int,
+        activation: FeedforwardActivation,
+        state_fallback: Callable[[torch.Tensor], torch.Tensor],
```
I like this, would make future code changes regarding state easier

Jegp commented:
Two items I'd like your feedback on:

* I'm considering dropping the linear part of the recurrent modules. Shouldn't it be up to the user to specify whether the input should be transformed?
* @cpehle had the idea that we could merge `SNNCell` and `FeedforwardSNNCell` into one layer, and parameterize the recurrence. It's pretty much the same for the user, and we could avoid confusing nomenclature (feedforward cell vs. just cell). Especially if we drop the linear part above.

Huizerd commented:
>
>     * I'm considering dropping the linear part of the recurrent modules. Shouldn't it be up to the user to specify whether the input should be transformed?
> 
>     * @cpehle had the idea that we could merge `SNNCell` and `FeedforwardSNNCell` into one layer, and parameterize the recurrence. It's pretty much the same for the user, and we could avoid confusing nomenclature (feedforward cell vs. just cell). Especially if we drop the linear part above.
>

Yes, I agree. Users would provide an input connection, e.g. `nn.Linear`, themselves, and optionally another `nn.Linear` (or something else) if they also want recurrency. So you would only need one base class, e.g. `SpikingCell` for pretty much any neuron type, feedforward or recurrent. There could then be another convenience class with the time loop like you had, e.g. `SpikingLayer`, such that we can have both the loop options you/I mentioned in one of the comments.

Huizerd commented:
Such that Norse only provides those things that people find most difficult about SNNs: dynamics and temporal loops

Jegp commented:
> Yes, I agree. Users would provide an input connection, e.g. `nn.Linear`, themselves, and optionally another `nn.Linear` (or something else) if they also want recurrency. So you would only need one base class, e.g. `SpikingCell` for pretty much any neuron type, feedforward or recurrent. There could then be another convenience class with the time loop like you had, e.g. `SpikingLayer`, such that we can have both the loop options you/I mentioned in one of the comments.

I actually quite like the idea of a `Linear` layer as recurrence. I initially thought to add a `recurrence: Union[bool, Tensor]` parameter because we want to describe three situations: no recurrence, recurrence with defaults, recurrence with specifics. But it's confusing and in your situation we could actually also add recurrent bias. It would then look something like this:
```python
LIFCell(2, 4, recurrence = None)            # 1. No recurrence (default?)
LIFCell(2, 4, recurrence = nn.Linear(4, 4)) # 2. Default recurrence
linear = nn.Linear(4, 4)                    # 3. Weighted recurrence
linear.weight = ...
LIFCell(2, 4, recurrence = linear)
```
The downside of this would be that the kaiming distribution of weights in the linear layer might not be optimal. But the upside is that we can just add any torch module to describe the recurrent dynamics.

I would also like to add a "temporal" layer, just to follow the Torch conventions. Such that we have `SNNCell`/`LIFCell` (w/o time) and `SNN`/ LIF` (with time). Does that sound 'digestible' to users?

Jegp commented:
Alright, I've given it another iteration. I liked the way you phrased the "responsibilities" of Norse @Huizerd: dynamics and temporal loops. At the same time I'm withdrawing on some of the more structural changes we talked about for two reasons: time and effort. Particularly, I think it requires a bit more careful thought to handle the recurrent parameters and find a proper way to implement the recurrence as modules. 

To avoid crippling complexity I'd suggest we settle on a user-facing class API and stick to the checklist I wrote in the initial PR. We can later adapt the interface as long as the main classes are in place.

Currently, I would suggest that we

1. Split non-recurrent neuron dynamics and time: `LIFCell` (without time) `LIF` (with time)
   * This aligns with PyTorch's API and is straight-forward to extend.
2. Split modules between recurrent and non-recurrent modules: `LIFRecurrent` vs. `LIF`
   * Getting the logic and parameters right is not something we should take too lightly I think
   * Furthermore, thinking about the recurrent dynamics, it might be better practice to split it up into two modules, since the recurrent neurons are logically separate. It also makes it easier to optimize stand-alone modules instead of dealing with logic inside the modules.

In sum, that would mean 4 modules for each neuron type: 
1. `LIFCell` (previous feedforward cell)
2. `LIFRecurrentCell` (did not exist before)
3. `LIF` (wraps the `LIFCell` in time)
4. `LIFRecurrent` (wraps the `LIFRecurrentCell` in time)

Huizerd commented:
Ok, after some more consideration it indeed seems more natural to have separate feedforward and recurrent classes. Now I'm thinking about the activation functions that are also different for feedforward and recurrent. Given that in the case of a recurrent neuron, the feedforward and recurrent inputs are just added together, could we have the following signature for an activation function:

```python
def activation(inputs: torch.Tensor, state: torch.Tensor, p: torch.nn.Module, dt: float) -> torch.Tensor:
    ...
    return outputs
```

and call it like this:

```python
# Feedforward
self.activation(inputs, state, self.p, self.dt)

# Recurrent option 1
inputs = self.input_weights @ inputs + self.recurrent_weights @ state.z
self.activation(inputs, state, self.p, self.dt)
# Recurrent option 2
self.rec_wrapper(self.activation, inputs, state, self.input_weights, self.recurrent_weights)

def rec_wrapper(activation: Callable, inputs: torch.Tensor, state: torch.Tensor, ff_weights: torch.Tensor, rec_weights: torch.Tensor) -> torch.Tensor:
    inputs = ff_weights @ inputs + rec_weights @ state.z
    return activation(inputs, state, self.p, self.dt)
```

Just to put this out there :)

Jegp commented:
I think that's a pretty good idea actually. And when you think about it, that doesn't change the user-facing API at all. It's just a matter of how we choose to structure the code internally!

If one were a complete OOP nerd, one could even imagine having a single class/function that implements the activation function, and then inherit into the other classes. I'm not certain it makes sense, but it could work. And it would remove the ludicrous code duplication errors from PyLint :facepalm:

Jegp commented:
Oh, and I deeply appreciate all the feedback @Huizerd btw. Thanks a lot!

Huizerd commented:
> If one were a complete OOP nerd, one could even imagine having a single class/function that implements the activation function, and then inherit into the other classes. I'm not certain it makes sense, but it could work. And it would remove the ludicrous code duplication errors from PyLint

Yes, that sounds like a good idea! And no problem, happy to help :)

cpehle commented on docs/source/conf.py:
```diff
@@ -42,16 +42,16 @@
     "sphinx.ext.githubpages",
     "sphinx_autodoc_typehints",
-    "sphinx_gallery.gen_gallery",
+    # "sphinx_gallery.gen_gallery",
```
remove entirely? I still like the idea of examples in the documentation, but maybe there is a better way.

Jegp replied:
I completely agree. At the same time I'm trying to really reduce the number of things we have to maintain. I'm also happy to leave it as-is. But as it stands, we only have one example which is a bit of a contrast to the tasks and notebooks, which pretty much have the same function.

cpehle replied:
Sure we should't keep stale content, I think it was the first thing I wrote to explain the library and it is kind of nice to have tutorial style content in the docs, but we should maybe stick to well maintained notebooks for now. One advantage of these examples is that they automatically get build and executed as part of the build process so it is easy to detect failures and they can be exported / opened as Jupyter notebooks.

cpehle replied:
In other words they should never be stale relative to the actual state of the library.

cpehle commented on docs/source/conf.py:
```diff
@@ -42,16 +42,16 @@
-    "examples_dirs": "../../norse/examples",  # path to your example scripts
-    "gallery_dirs": "auto_examples",  # path to where to save gallery generated output
-}
+# sphinx_gallery_conf = {
+#     "examples_dirs": "../../norse/examples",  # path to your example scripts
+#     "gallery_dirs": "auto_examples",  # path to where to save gallery generated output
+# }
```
See above

cpehle commented on README.md:
```diff
@@ -116,19 +116,20 @@ This classifier is a taken from our [tutorial on training a spiking MNIST classi
 model = SequentialState(
     nn.Conv2d(3, 20, 5, 1),      # Convolve from 3 -> 20 channels
-    LIFFeedForwardCell(),        # Spiking activation layer
+    LIFCell(),                   # Spiking activation layer
```
Inconsistent with what you are writing in getting started below.

Jegp replied:
Good catch. I haven't fully looked through the docs. I'll go through it now

cpehle commented on docs/source/started.rst:
```diff
@@ -73,7 +73,7 @@ Luckily, we offer our own `SequentialState <https://norse.github.io/norse/auto_a
     model = SequentialState(
         torch.nn.Linear(10, 5),
-        norse.LIFFeedForwardCell(),
+        norse.LIF(),
```
Is it LIFCell or LIF? See above.

Jegp replied:
I updated the documentation (mainly getting started and working with pages). It's actually easier to explain, which I guess is a good sign.

cpehle replied:
Yeah I think that is a good sign!

cpehle commented on norse/torch/functional/lif.py:
```diff
@@ -22,13 +22,24 @@ class LIFParameters(NamedTuple):
-    v_reset: torch.Tensor = torch.as_tensor(0.0)
-    method: str = "super"
-    alpha: float = torch.as_tensor(100.0)
+    def __init__(
+        self,
+        tau_syn_inv: torch.Tensor = torch.as_tensor(1.0 / 5e-3),
+        tau_mem_inv: torch.Tensor = torch.as_tensor(1.0 / 1e-2),
+        v_leak: torch.Tensor = torch.as_tensor(0.0),
+        v_th: torch.Tensor = torch.as_tensor(1.0),
+        v_reset: torch.Tensor = torch.as_tensor(0.0),
+        method: str = "super",
+        alpha: float = torch.as_tensor(100.0),
+    ):
+        super().__init__()
+        self.tau_syn_inv = tau_syn_inv
+        self.tau_mem_inv = tau_mem_inv
+        self.v_leak = v_leak
+        self.v_th = v_th
+        self.v_reset = v_reset
+        self.method = method
+        self.alpha = alpha
```
I thought in this changeset we wouldn't switch over to torch.nn.Module for the parameters.

Jegp replied:
Indeed, this is a remnant of the previous example. Thanks for pointing this out!

cpehle reviewed (changes requested):
I think we can merge these changes as is pending the few minor changes. We should then iterate based on your discussion in this pull request + issue on the outstanding changes. Overall great work and I think this is a good improvement!

cpehle reviewed (approved):
LGTM for now, should probably iterate more to address the other points raised in the discussion. But this is a great start!

Jegp commented:
Great. I also like where we're heading, but there could be a few things I missed.

I'll make sure the static code analysis and coverage looks ok and then we can merge. Thanks for your feedback @cpehle and @Huizerd!
